keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,875 @@
+In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
+<<[/UNRESOLVED]>>
+* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
+* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
+* For Delete, this means that the oldObject must match the field and label selector.
+ 
+We do not allow field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
+field selector and use a List to get equivalent functionality.
+
+### SubjectAccessReview Changes
+
+SubjectAccessReview is used for two purposes:
+1. Authorization webhook calls from the kube-apiserver to a webhook.
+   This usage likely benefits from a serialization with `[]Requirement`.
+2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
+   This usage likely benefits from a serialization that matches the query parameter.
+
+Their needs best met with two different serialization (see user stories)
+
+```go
+
+type SubjectAccessReviewSpec struct {
+	ResourceAttributes *ResourceAttributes
+}
+
+type ResourceAttributes struct {
+    // If the SubjectAccessReview specified verb,resource tuple does not honor field selection, the fieldSelector is ignored.
+	FieldSelector *FieldSelectorAttributes
+
+    // If the SubjectAccessReview specified verb,resource tuple does not honor label selection, the labelSelector is ignored.
+	LabelSelector *LabelSelectorAttributes
+}
+
+// FieldSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// * If rawSelector is empty and requirements are empty, the request is not limited.
+// * If rawSelector is present and requirements are empty, the request is not limited.
+// * If rawSelector is empty and requirements are present, the requirements should be honored
+// * If rawSelector is present and requirements are present, the request is invalid.
+// Webhook authors are encouraged to
+// * ensure rawSelector and requirements are not both set
+// * consider the requirements field if set
+// * not try to parse or consider the rawSelector field if set.
+//   This avoid this is to avoid another CVE-2022-2880 (i.e. getting different systems to agree on how exactly to parse
+//   a query is not something we want), see https://www.oxeye.io/resources/golang-parameter-smuggling-attack for more details.
+// For the kube-apiserver:
+// * If rawSelector is empty and requirements are empty, the request is not limited.
+// * If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// * If rawSelector is empty and requirements are present, the requirements should be honored
```

Comment:
from https://github.com/kubernetes/enhancements/pull/4600/files#r1593019928

> client -> kube-apiserver SAR endpoint: if the kube-apiserver would not honor the selectors, they are removed prior to calling the authorizer.

for the "rawSelector is present" and "requirements are present" cases, is there a verb constraint as well? The verb must be one kube-apiserver would honor the selector for in order for the request to be limited?

(same question on label selector docs)

Reply:
> for the "rawSelector is present" and "requirements are present" cases, is there a verb constraint as well? The verb must be one kube-apiserver would honor the selector for in order for the request to be limited?

Doing so would preclude a N+1 aggregated apiserver that allows additional verbs from using in-cluster authorization if the kube-apiserver doesn't also know about that verb.  Allowing any verb keeps this possible and only impacts the client.

Reply:
I guess I don't understand how to reconcile that with the text at https://github.com/kubernetes/enhancements/pull/4600/files#r1593019928

Reply:
> Does kube-apiserver honor selector requirements in a SAR for a verb that doesn't support selectors or not?

When receiving a request, the request info resolver only adds field and label selector information to authorization attributes when the kube-apiserver honors those selectors for the verb being used.

When handling a SAR, no verb interpretation is done prior to adding the field and label selectors.

This allows skew in the cluster while ensuring the kube-apiserver request handling only considers field and label selectors for certain verbs.

Reply:
> When receiving a request, the request info resolver only adds field and label selector information to authorization attributes when the kube-apiserver honors those selectors for the verb being used.
>
> When handling a SAR, no verb interpretation is done prior to adding the field and label selectors.
>
> This allows skew in the cluster while ensuring the kube-apiserver request handling only considers field and label selectors for certain verbs.

Ok, I think I agree with that. Sweep the KEP / API docs to make that consistent and clear.

For example, the text in the KEP still says "SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and label selector attributes removed prior to checking authorization." which I think is directly opposed to "When handling a SAR, no verb interpretation is done prior to adding the field and label selectors"

Reply:
Updated.  Caught a couple others too.